### PR TITLE
Update WebConsoleConfiguration

### PR DESCRIPTION
**Pull request by spadgett**
* Add new stanza for cluster config like master and metrics URLs
* Add new stanza for extensions config
* Remove obsolete extension properties
* Rename script and stylesheet properties since they're now URLs
* Add inactivityTimeoutMinutes option

/assign @deads2k 
/cc @liggitt @jwforres

### CODE REVIEW

**Comment by spadgett** [comment]
I have not added config for ClusterResourceOverrides yet

**Comment by deads2k** [comment]
@spadgett something is red.  For ease of review can you split the manual changes from the generated ones?

File: webconsole/v1/types.go [hunk 1]
```diff
@@ -10,16 +10,31 @@ import (
...
-
-	// PublicURL is where you can find the asset server (TODO do we really need this?)
-	PublicURL string `json:"publicURL" protobuf:"bytes,2,opt,name=publicURL"`
+	// PublicURL is where you can find the web console server (TODO do we really need this?)
+	PublicURL string `json:"publicURL" protobuf:"bytes,1,opt,name=publicURL"`
 
 	// LogoutURL is an optional, absolute URL to redirect web browsers to after logging out of the web console.
 	// If not specified, the built-in logout page is shown.
-	LogoutURL string `json:"logoutURL" protobuf:"bytes,3,opt,name=logoutURL"`
+	LogoutURL string `json:"logoutURL" protobuf:"bytes,2,opt,name=logoutURL"`
```
Comment (deads2k) [inline webconsole/v1/types.go]: This and the inactivity look related to "authentication"

File: webconsole/v1/types.go [hunk 2]
```diff
@@ -10,16 +10,31 @@ import (
...
+
+	// InactivityTimeoutMinutes is the number of minutes of inactivity before you are automatically logged out of
+	// the web console (optional). If set to 0, inactivity timeout is disabled.
+	InactivityTimeoutMinutes int64 `json:"inactivityTimeoutMinutes" protobuf:"varint,3,opt,name=inactivityTimeoutMinutes"`
+
+	// ClusterConfiguration holds information the web console needs to talk to the cluster such as master public URL
+	// and metrics public URL.
+	ClusterConfiguration []ClusterConfiguration `json:"clusterConfiguration" protobuf:"bytes,4,rep,name=clusterConfiguration"`
```
Comment (deads2k) [inline webconsole/v1/types.go]: I thought the console only handled one cluster.

File: webconsole/v1/types.go [hunk 3]
```diff
@@ -10,16 +10,31 @@ import (
 type WebConsoleConfiguration struct {
 	metav1.TypeMeta `json:",inline"`
 
-	// ServingInfo is the HTTP serving information for these assets
-	ServingInfo HTTPServingInfo `json:"servingInfo" protobuf:"bytes,1,opt,name=servingInfo"`
-
-	// PublicURL is where you can find the asset server (TODO do we really need this?)
-	PublicURL string `json:"publicURL" protobuf:"bytes,2,opt,name=publicURL"`
+	// PublicURL is where you can find the web console server (TODO do we really need this?)
+	PublicURL string `json:"publicURL" protobuf:"bytes,1,opt,name=publicURL"`
```
Comment (deads2k) [inline webconsole/v1/types.go]: under ClusterConfiguration?

File: webconsole/v1/types.go [hunk 4]
```diff
@@ -10,16 +10,31 @@ import (
...
+	// ServingInfo is the HTTP serving information for these assets
+	ServingInfo HTTPServingInfo `json:"servingInfo" protobuf:"bytes,5,opt,name=servingInfo"`
+
+	// Extensions define custom scripts, stylesheets, and properties used for web console customization
+	Extensions []ExtensionsConfiguration `json:"extensions" protobuf:"bytes,6,rep,name=extensions"`
+}
+
+// ClusterConfiguration holds information the web console needs to talk to the cluster such as master public URL and
+// metrics public URL.
+type ClusterConfiguration struct {
```
Comment (deads2k) [inline webconsole/v1/types.go]: You're envisioning CRO under this?

File: webconsole/v1/types.go [hunk 5]
```diff
@@ -28,40 +43,21 @@ type WebConsoleConfiguration struct {
...
-	// the Web Console loads
-	ExtensionStylesheets []string `json:"extensionStylesheets" protobuf:"bytes,9,rep,name=extensionStylesheets"`
-
-	// Extensions are files to serve from the asset server filesystem under a subcontext
-	Extensions []AssetExtensionsConfig `json:"extensions" protobuf:"bytes,10,rep,name=extensions"`
+// ExtensionsConfiguration holds custom script, stylesheets, and properties used for web console customization
+type ExtensionsConfiguration struct {
+	// ExtensionScriptURLs are URLs to load as scripts when the Web Console loads. The URLs must be accessible from
+	// the browser.
+	ExtensionScriptURLs []string `json:"extensionScriptURLs" protobuf:"bytes,1,rep,name=extensionScriptURLs"`
```
Comment (deads2k) [inline webconsole/v1/types.go]: Interesting.  No need for a map anymore?

File: webconsole/v1/types.go [hunk 6]
```diff
@@ -10,16 +10,31 @@ import (
...
+	// InactivityTimeoutMinutes is the number of minutes of inactivity before you are automatically logged out of
+	// the web console (optional). If set to 0, inactivity timeout is disabled.
+	InactivityTimeoutMinutes int64 `json:"inactivityTimeoutMinutes" protobuf:"varint,3,opt,name=inactivityTimeoutMinutes"`
+
+	// ClusterConfiguration holds information the web console needs to talk to the cluster such as master public URL
+	// and metrics public URL.
+	ClusterConfiguration []ClusterConfiguration `json:"clusterConfiguration" protobuf:"bytes,4,rep,name=clusterConfiguration"`
 
+	// ServingInfo is the HTTP serving information for these assets
+	ServingInfo HTTPServingInfo `json:"servingInfo" protobuf:"bytes,5,opt,name=servingInfo"`
```
Comment (deads2k) [inline webconsole/v1/types.go]: nit: top of struct.

**Comment by spadgett** [comment]
I need help fixing the travis error. `make generate` shows no changes. Could it be different version of protobuf?

File: webconsole/v1/types.go (hunk 4 above)
Comment (spadgett) [reply webconsole/v1/types.go]: Yes

**Comment by deads2k** [comment]
> I need help fixing the travis error. make generate shows no changes. Could it be different version of protobuf?

Once we get hte types pinned down, I'll pull it and generate

File: webconsole/v1/types.go (hunk 5 above)
Comment (spadgett) [reply webconsole/v1/types.go]: There's still extension properties, but the old `extensions` stanza from asset config is no longer supported

File: webconsole/v1/types.go [hunk 7]
```diff
@@ -13,55 +13,51 @@ type WebConsoleConfiguration struct {
...
+	Extensions ExtensionsConfiguration `json:"extensions" protobuf:"bytes,5,rep,name=extensions"`
+}
 
-	// MetricsPublicURL is the public endpoint for metrics (optional)
-	MetricsPublicURL string `json:"metricsPublicURL" protobuf:"bytes,6,opt,name=metricsPublicURL"`
+// ClusterInfo holds information the web console needs to talk to the cluster such as master public URL and
+// metrics public URL
+type ClusterInfo struct {
+	// MasterPublicURL is how the web console can access the OpenShift v1 server
+	MasterPublicURL string `j4on:"masterPublicURL" protobuf:"bytes,1,opt,name=masterPublicURL"`
```
Comment (deads2k) [inline webconsole/v1/types.go]: j4on?

Comment (deads2k) [reply webconsole/v1/types.go]: > j4on?

@sttts is there a linter of a test we have that would catch stuff like this?

File: webconsole/v1/types.go [hunk 8]
```diff
@@ -13,55 +13,51 @@ type WebConsoleConfiguration struct {
...
 
-	// ExtensionScripts are file paths on the asset server files to load as scripts when the Web
-	// Console loads
-	ExtensionScripts []string `json:"extensionScripts" protobuf:"bytes,7,rep,name=extensionScripts"`
+	// LogoutURL is an optional, absolute URL to redirect web browsers to after logging out of the web console.
+	// If not specified, the built-in logout page is shown.
+	LogoutURL string `json:"logoutURL" protobuf:"bytes,2,opt,name=logoutURL"`
```
Comment (deads2k) [inline webconsole/v1/types.go]: `LogoutPublicURL` ?

**Comment by spadgett** [comment]
Updated

**Comment by spadgett** [comment]
/hold

some debate still on the CRO flag

**Comment by spadgett** [comment]
/hold cancel

@jwforres discussed more, and we think this is good for now. We can always revisit the CRO flags later before release since they have no install or cluster up impact.

File: webconsole/v1/types.go (hunk 7 above)
Comment (sttts) [reply webconsole/v1/types.go]: j4on is valid tag name, isn't it?

File: webconsole/v1/types.go [hunk 9]
```diff
@@ -13,55 +13,59 @@ type WebConsoleConfiguration struct {
...
-
-	// ExtensionStylesheets are file paths on the asset server files to load as stylesheets when
-	// the Web Console loads
-	ExtensionStylesheets []string `json:"extensionStylesheets" protobuf:"bytes,9,rep,name=extensionStylesheets"`
-
-	// Extensions are files to serve from the asset server filesystem under a subcontext
-	Extensions []AssetExtensionsConfig `json:"extensions" protobuf:"bytes,10,rep,name=extensions"`
+	// ClusterResourceOverridesEnabled indicates that the cluster is configured for overcommit and the web console
+	// should hide request and limit fields that will be overridden in its editors
+	ClusterResourceOverridesEnabled bool `json:"clusterResourceOverridesEnabled" protobuf:"varint,6,opt,name=clusterResourceOverridesEnabled"`
```
Comment (deads2k) [inline webconsole/v1/types.go]: You don't need specific values?

File: webconsole/v1/types.go (hunk 7 above)
Comment (deads2k) [reply webconsole/v1/types.go]: > j4on is valid tag name, isn't it?

Valid tag name, but there should be a json tag for every external field that doesn't explicitly opt out.

**Comment by spadgett** [comment]
Removed the CRO flag until we know for sure what we want.

**Pull request merged**